### PR TITLE
Linting Rules Additions - Proposal

### PR DESCRIPTION
**Pull request by meghfossa**
# Overview

Received some great feedback on styling of haskell code over last week. I would like to embed those suggestions into linter. Leaving this as a draft before making changes to existing code for style corrections (per this pr)

## Acceptance criteria

- [x] Adds linting rules for `void` over `()`
- [x] Adds linting rules for preference of `pure` over `return`
- [x] Adds linting rules for preference of `(<|>)` over `asum`, when considering only two items
- [x] Adds linting rules for preference of `<$>` over `<&>`
- [x] Adds linting rules for preference of complete name for qualified import

## Testing plan

N/A

## Risks

N/A

## References

This responds to https://github.com/fossas/team-analysis/issues/671

## Checklist

- [x] I added tests for this PR's change (or confirmed tests are not viable).
- [x] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [x] I updated `Changelog.md` if this change is externally facing. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.
- [x] I linked this PR to any referenced GitHub issues, if they exist.

### CODE REVIEW

**Review by liftM — APPROVED** [review]
LGTM. Let's get this in sooner rather than later, so we can iterate on it.

@scruffystuffs our CI's automatic `hlint` will pick up this configuration and enforce it, right?

**Comment by scruffystuffs** [comment]
@liftM Looks like it's already doing it.

File: .hlint.yaml [hunk 1]
```diff
@@ -1,4 +1,18 @@
...
+
+# Import Preferences
+- modules:
+  - {name: Data.Set, as: Set, message: "Use complete name for qualified import of Set"}
+  - {name: Data.Map, as: Map, message: "Use complete name for qualified import of Map"}
+  - {name: Data.Text, as: Text, message: "Use complete name for qualified import of Text"}
+
+# Styling Preferences
+- suggest: {lhs: "asum [x, y]", rhs: x <|> y}
+- suggest: {lhs: return x, rhs: pure x}
```
Comment (scruffystuffs) [inline .hlint.yaml:14]: I think this is causing a rewrite from `return $ f a` to `pure (f a)`.  I think rule can be rewritten to fix that.
```suggestion
- suggest: {lhs: return, rhs: pure}
```

File: .hlint.yaml [hunk 2]
```diff
@@ -1,4 +1,18 @@
...
+- suggest: {lhs: () <$ x, rhs: Control.Monad.void x}
+- suggest: {lhs: x <&> f, rhs: f <$> x}
+- suggest: {lhs: Data.Text.pack x, rhs: Data.String.Conversion.toText x, note: "For consistency use, Data.String.Conversion module instead!"}
+- suggest: {lhs: Data.Text.unpack x, rhs: Data.String.Conversion.toString x, note: "For consistency use, Data.String.Conversion module instead!"}
```
Comment (scruffystuffs) [inline .hlint.yaml:18]: Don't justify your choices in the warnings.  Justifications should be in a style guide, and they can be left out until a style guide is ready, as long as we **internally** agree on correct decisions.
```suggestion
- suggest: {lhs: Data.Text.pack x, rhs: Data.String.Conversion.toText x}
- suggest: {lhs: Data.Text.unpack x, rhs: Data.String.Conversion.toString x}
```

**Review by scruffystuffs — COMMENTED** [review]
Let's add a rule for `toFilePath` becoming `toString`.  Let's also make sure that file doesn't get those specific lint errors, but it does get others, I think we just need to give them a name and add specific no-lint comments to the file.

I also wonder if we can automatically make a `toText` instance if there's a `toString` instance.  This would be useful for simplifying `toText . toString` into just `toText`.  I have a feeling that it won't work, but it's worth a shot.

**Comment by scruffystuffs** [comment]
We should additionally add rules to prevent the use of `error` and `undefined`.  These can be useful in development, but dangerous in production.

**Comment by liftM** [comment]
https://github.com/kowainik/relude is an alternative Prelude that I frequently use in my side projects that also comes with a bunch of great hlint rules we can steal (when relevant): https://github.com/kowainik/relude/blob/main/.hlint.yaml

> I also wonder if we can automatically make a `toText` instance if there's a `toString` instance. This would be useful for simplifying `toText . toString` into just `toText`. I have a feeling that it won't work, but it's worth a shot.

```hs
instance (ToString t) => ToText t where
  toText = T.pack . toString
```

**Comment by scruffystuffs** [comment]
@liftM does that compile?  I still haven't checked, but I think it might give us an infinite check or undecidable instance error?

**Review by scruffystuffs — APPROVED** [review]
I didn't check every file, but I went through a bunch of them.  Looks good.  Left a comment about `error` and `undefined`.

File: src/App/Fossa/ProjectInference.hs [hunk 3]
```diff
@@ -65,7 +65,7 @@ inferProjectDefault dir = sendIO $ do
   let name = FP.dropTrailingPathSeparator (fromRelDir (dirname dir))
   time <- floor <$> getPOSIXTime :: IO Int
 
-  pure (InferredProject (T.pack name) (T.pack (show time)) Nothing)
+  pure (InferredProject (toText name) (toText (show time)) Nothing)
```
[the remark this answers is not in the document]
Comment (meghfossa) [reply src/App/Fossa/ProjectInference.hs:68]: We can - and ignore bracket hint when apt - https://github.com/ndmitchell/hlint#ignoring-hints. Although it does not allow disabling at a line, only at function or module level.

[the remark this answers is not in the document]
Comment (scruffystuffs) [reply src/App/Fossa/ProjectInference.hs:68]: I'm thinking we just use it as-is.   Any time I see `((` or `))` I become sad.  Haskell isn't a Lisp.

**Pull request merged**